File: ios/Photo/Rendering/CropGeometry.swift
import CoreGraphics

/// Resolution-independent constraints; mirrors Android CropGeometry.
enum CropGeometry {
  enum Handle { case move, top, right, bottom, left, topLeft, topRight, bottomLeft, bottomRight }
  static func fillScale(width: CGFloat, height: CGFloat, degrees: CGFloat) -> CGFloat {
    let radians = degrees * .pi / 180
    let c = abs(cos(radians)), s = abs(sin(radians))
    return max(c + height / width * s, c + width / height * s)
  }
  static func fit(_ bounds: CGRect, ratio: CGFloat?) -> CGRect {
    guard let ratio, ratio > 0 else { return bounds }
    let w = min(bounds.width, bounds.height * ratio), h = min(bounds.width / ratio, bounds.height)
    return CGRect(x: bounds.midX - w / 2, y: bounds.midY - h / 2, width: w, height: h)
  }
  static func resize(_ r: CGRect, bounds: CGRect, handle: Handle, dx: CGFloat, dy: CGFloat, ratio: CGFloat?, minimum: CGFloat) -> CGRect {
    if handle == .move {
      return r.offsetBy(dx: max(bounds.minX-r.minX, min(bounds.maxX-r.maxX, dx)), dy: max(bounds.minY-r.minY, min(bounds.maxY-r.maxY, dy)))
    }
    let left = [Handle.left, .topLeft, .bottomLeft].contains(handle)
    let right = [Handle.right, .topRight, .bottomRight].contains(handle)
    let top = [Handle.top, .topLeft, .topRight].contains(handle)
    let bottom = [Handle.bottom, .bottomLeft, .bottomRight].contains(handle)
    var l = r.minX + (left ? dx : 0), t = r.minY + (top ? dy : 0)
    var rr = r.maxX + (right ? dx : 0), b = r.maxY + (bottom ? dy : 0)
    if let ratio, ratio > 0 {
      if (left || right) && (top || bottom) {
        let dw = left ? -dx : dx, dh = top ? -dy : dy
        let w = r.width + (abs(dw) > abs(dh * ratio) ? dw : dh * ratio)
        if left { l = r.maxX - w } else { rr = r.minX + w }
        if top { t = r.maxY - w / ratio } else { b = r.minY + w / ratio }
      } else if left || right {
        let h = (rr-l)/ratio; t = r.midY-h/2; b = t+h
      } else {
        let w = (b-t)*ratio; l = r.midX-w/2; rr = l+w
      }
    }
    var fraction: CGFloat = 1
    func limit(_ before: CGFloat, _ after: CGFloat, _ minimum: CGFloat) {
      if after < minimum && after < before { fraction = min(fraction, max(0, min(1, (before-minimum)/(before-after)))) }
    }
    limit(r.minX,l,bounds.minX); limit(r.minY,t,bounds.minY)
    limit(-r.maxX,-rr,-bounds.maxX); limit(-r.maxY,-b,-bounds.maxY)
    limit(r.width,rr-l,min(minimum,r.width)); limit(r.height,b-t,min(minimum,r.height))
    let x = r.minX+(l-r.minX)*fraction, y = r.minY+(t-r.minY)*fraction
    return CGRect(x: x, y: y, width: r.maxX+(rr-r.maxX)*fraction-x, height: r.maxY+(b-r.maxY)*fraction-y)
  }
}

File: android/src/main/java/com/photovideoeditor/photo/render/CropGeometry.kt
package com.photovideoeditor.photo.render

import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin

/** Resolution-independent crop constraints shared by gestures and rendering. */
object CropGeometry {
  data class Rect(val left: Float, val top: Float, val right: Float, val bottom: Float) {
    val width get() = right - left
    val height get() = bottom - top
  }
  enum class Handle { MOVE, TOP, RIGHT, BOTTOM, LEFT, TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT }

  fun fillScale(width: Float, height: Float, degrees: Float): Float {
    val radians = Math.toRadians(degrees.toDouble())
    val c = abs(cos(radians)).toFloat(); val s = abs(sin(radians)).toFloat()
    return maxOf(c + height / width * s, c + width / height * s)
  }

  fun fit(bounds: Rect, ratio: Float?): Rect {
    if (ratio == null || ratio <= 0) return bounds
    val width = minOf(bounds.width, bounds.height * ratio)
    val height = width / ratio
    val x = (bounds.left + bounds.right - width) / 2
    val y = (bounds.top + bounds.bottom - height) / 2
    return Rect(x, y, x + width, y + height)
  }

  /** Keeps the opposite edge/corner anchored; clamps motion instead of shifting a resized box. */
  fun resize(r: Rect, bounds: Rect, handle: Handle, dx: Float, dy: Float, ratio: Float?, minimum: Float): Rect {
    if (handle == Handle.MOVE) {
      val x = dx.coerceIn(bounds.left - r.left, bounds.right - r.right)
      val y = dy.coerceIn(bounds.top - r.top, bounds.bottom - r.bottom)
      return Rect(r.left + x, r.top + y, r.right + x, r.bottom + y)
    }
    val left = handle in listOf(Handle.LEFT, Handle.TOP_LEFT, Handle.BOTTOM_LEFT)
    val right = handle in listOf(Handle.RIGHT, Handle.TOP_RIGHT, Handle.BOTTOM_RIGHT)
    val top = handle in listOf(Handle.TOP, Handle.TOP_LEFT, Handle.TOP_RIGHT)
    val bottom = handle in listOf(Handle.BOTTOM, Handle.BOTTOM_LEFT, Handle.BOTTOM_RIGHT)
    var l = r.left + if (left) dx else 0f
    var t = r.top + if (top) dy else 0f
    var rr = r.right + if (right) dx else 0f
    var b = r.bottom + if (bottom) dy else 0f
    if (ratio != null && ratio > 0) {
      if ((left || right) && (top || bottom)) {
        val deltaW = if (left) -dx else dx
        val deltaH = if (top) -dy else dy
        val width = r.width + if (abs(deltaW) > abs(deltaH * ratio)) deltaW else deltaH * ratio
        if (left) l = r.right - width else rr = r.left + width
        if (top) t = r.bottom - width / ratio else b = r.top + width / ratio
      } else if (left || right) {
        val height = (rr - l) / ratio
        t = (r.top + r.bottom - height) / 2; b = t + height
      } else {
        val width = (b - t) * ratio
        l = (r.left + r.right - width) / 2; rr = l + width
      }
    }
    // All constraints are linear along the drag; interpolation preserves the ratio and anchor.
    var fraction = 1f
    fun limit(before: Float, after: Float, minimumValue: Float) {
      if (after < minimumValue && after < before) fraction = minOf(fraction, ((before - minimumValue) / (before - after)).coerceIn(0f, 1f))
    }
    limit(r.left, l, bounds.left); limit(r.top, t, bounds.top)
    limit(-r.right, -rr, -bounds.right); limit(-r.bottom, -b, -bounds.bottom)
    limit(r.width, rr - l, minOf(minimum, r.width)); limit(r.height, b - t, minOf(minimum, r.height))
    return Rect(r.left + (l - r.left) * fraction, r.top + (t - r.top) * fraction,
      r.right + (rr - r.right) * fraction, r.bottom + (b - r.bottom) * fraction)
  }
}
